src/_posts/2020-03-24-project-maturity-model.md

```diff
@@ -0,0 +1,153 @@
+
+One of our internal services had changed it's contract and one of the services could not cope with the new shape of the data. This did not take long to solve or isolate given the now improving diagrams.
+
+### Level 3 Defined
+
+We now started to reduce the most frequent group of errors whenever we had a quiet time. Datadog has a great option to detect patterns in the logs and create groups out of them. 
+These form obvious targets for reducing the errors. We started adding alerts for specific problems to ensure that we know about a problem before the users tell us.
+
+We were still occasionally getting problems reported just before hometime on a Friday. Solving these is much quicker now.
+
+
+## Third Crisis
+Our QA while testing a change noticed a message on a database dashboard. The database service that we were using for eight of the services is being decomissioned in 6 months time.
+We fairly quickly find that four of the databases are actually only used for locking and were able to replace them with AWS Lamdas. The remaining four we eventually migrated.
+
+The diagrams are getting more accurate by this point.
+
+## Further Improvements
+
+We added integration tests to check that the access codes that we needed for each remote service were still accurate. We are integrating with over 100 almost identical services and they are undergoing an upgrade program! This means that the tests will be broken first thing on Monday morning, allowing us to fix things before our users need the data.
+
+One of the services used AWS S3 to store messages that could not be parsed. Expanding this to other parts of the service greatly assisted fixing (or at least quieting) some strange errors.
+
+One of the services was responsible for querying an external service (which had over 100 instances). At any given time some of these were broken (2% to 5%, but not always the same one). For a while we were moving the broken feeds to use a less effictive backup service manually. This required us to remember to restore the feeds once they had recovered.
+
+The first attempt to standardise this involved having a test check each feed each day and report those that were broken. This was good at spotting broken items, but made fixing them labour intensive.
+
+The second attempt was better: use railway oriented programming. This is a technique which models the success and error conditions as two parallel railway tracks. An error shunts the train onto the error track. If it remained there then this would be normal error handling. The railway trick is to have corrective processes that can restore the train to the main track. In our specific case this meant changing the error handler of the primary feed to call the secondary feed should an error occour. This allowed us to cater for unreliable feeds for the small price of one service being able to call another. This small change eliminated 30% of the daily errors!
```

Comment:
Change sentence: In our specific case, this meant changing the error handler of the primary feed to call the secondary feed should an error occur.